core/trace/dt_span_exporter.go

```diff
@@ -138,9 +125,44 @@ func (e *dtSpanExporterImpl) export(ctx context.Context, t exportType, spans dtS
 		return errors.New("unexpected response code: " + strconv.Itoa(resp.StatusCode))
 	}
 
+	e.logger.Debug("Export successful")
 	return nil
 }
 
+func (e *dtSpanExporterImpl) exportWithRetry(ctx context.Context, t exportType, serializedSpans []byte) (*http.Response, error) {
+	req, err := e.createExportRequest(ctx, serializedSpans)
+	if err != nil {
+		return nil, err
+	}
+	exportStartTime := time.Now()
+	resp, err := e.performHttpRequest(req, t)
+	exportDuration := time.Since(exportStartTime)
+
+	// Retry once if the connection was reset by the backend and the request took less than cConnectionResetTimeLimit.
+	if errors.Is(err, io.EOF) {
+		if exportDuration < cConnectionResetTimeLimit {
+			e.logger.Warn("Got EOF during export, retrying once")
+			req, err := e.createExportRequest(ctx, serializedSpans)
+			if err != nil {
+				return nil, err
+			}
+			return e.performHttpRequest(req, t)
+		} else {
+			e.logger.Warnf("Got EOF during export, but request took too long (%v), not retrying", exportDuration)
+		}
+	}
+	return resp, err
+}
+
+func (e *dtSpanExporterImpl) createExportRequest(ctx context.Context, serializedSpans []byte) (*http.Request, error) {
+	body := bytes.NewBuffer(serializedSpans)
```

Comment:
Docs say: https://pkg.go.dev/bytes#NewBuffer
> NewBuffer creates and initializes a new Buffer using buf as its initial contents. The new Buffer takes ownership of buf, and the caller should not use buf after this call.

But in a retry, we seem to violate this. Could we instead reuse the whole buffer? Or maybe this warning is only relevant if somebody uses the buffer to write?

Reply:
Correct, I think this warning only applies to situations where the buffer is written to, which modifies the underlying byte array, like in this demonstration where it may cause "unexpected" problems: https://go.dev/play/p/ERqoY7_PMG

While it should not cause any issues in this case, I think converting this to use [`bytes.NewReader()`](https://pkg.go.dev/bytes#NewReader) instead makes more sense, which is read-only and can be reused via its `Reset()` method. We don't need the write capabilities of the `Buffer`.